p2p/protocol/holepunch/coordination.go

```diff
@@ -0,0 +1,285 @@
+				log.Infof("direct connection to peer %s successful, no need for a hole punch", rp.Pretty())
+				return
+			}
+			break
+		}
+	}
+
+	// hole punch
+	s, err := hs.host.NewStream(hs.ctx, rp, protocol)
+	if err != nil {
+		log.Errorf("initiator failed to open hole punching stream to peer %s, err: %s", rp, err)
+		return
+	}
+	_ = s.SetDeadline(time.Now().Add(holePunchTimeout))
+	w := protoio.NewDelimitedWriter(s)
+
+	// send a CONNECT and start RTT measurement
+	msg := new(holepunch_pb.HolePunch)
+	msg.Type = holepunch_pb.HolePunch_CONNECT.Enum()
+	msg.ObsAddrs = addrsToBytes(hs.ids.OwnObservedAddrs())
+	if err := w.WriteMsg(msg); err != nil {
+		s.Reset()
+		log.Errorf("failed to send hole punch CONNECT, err: %s", err)
+		return
+	}
+	tstart := time.Now()
+
+	// wait for a CONNECT message from the remote peer
+	rd := protoio.NewDelimitedReader(s, maxMsgSize)
+	msg.Reset()
+	if err := rd.ReadMsg(msg); err != nil {
+		s.Reset()
+		log.Errorf("failed to read connect message from remote peer, err: %s", err)
+		return
+	}
+	if msg.GetType() != holepunch_pb.HolePunch_CONNECT {
+		s.Reset()
+		return
+	}
+	obsRemote := addrsFromBytes(msg.ObsAddrs)
+	rtt := time.Since(tstart)
+
+	// send a SYNC message and attempt a direct connect after half the RTT
+	msg.Reset()
+	msg.Type = holepunch_pb.HolePunch_SYNC.Enum()
+	if err := w.WriteMsg(msg); err != nil {
+		s.Reset()
+		log.Errorf("failed to send SYNC message for hole punching, err: %s", err)
+		return
+	}
+	defer s.Close()
+
+	synTime := time.Duration(rtt.Milliseconds()/2) * time.Millisecond
+
+	// wait for sync to reach the other peer and then punch a hole for it in our NAT
+	// by attempting a connect to it.
+	select {
+	case <-time.After(synTime):
+		pi := peer.AddrInfo{
+			ID:    rp,
+			Addrs: obsRemote,
+		}
+		hs.holePunchConnectWithBackoff(pi)
+
+	case <-hs.ctx.Done():
+		return
+	}
+}
+
+func (hs *HolePunchService) handleNewStream(s network.Stream) {
+	_ = s.SetDeadline(time.Now().Add(holePunchTimeout))
+	rp := s.Conn().RemotePeer()
+	wr := protoio.NewDelimitedWriter(s)
+	rd := protoio.NewDelimitedReader(s, maxMsgSize)
+
+	// Read Connect message
+	msg := new(holepunch_pb.HolePunch)
+	if err := rd.ReadMsg(msg); err != nil {
+		s.Reset()
+		return
+	}
+	if msg.GetType() != holepunch_pb.HolePunch_CONNECT {
+		s.Reset()
+		return
+	}
+	obsDial := addrsFromBytes(msg.ObsAddrs)
+
+	// Write CONNECT message
+	msg.Reset()
+	msg.Type = holepunch_pb.HolePunch_CONNECT.Enum()
+	msg.ObsAddrs = addrsToBytes(hs.ids.OwnObservedAddrs())
+	if err := wr.WriteMsg(msg); err != nil {
+		s.Reset()
+		return
+	}
+
+	// Read SYNC message
+	msg.Reset()
+	if err := rd.ReadMsg(msg); err != nil {
+		s.Reset()
+		return
+	}
+	if msg.GetType() != holepunch_pb.HolePunch_SYNC {
+		s.Reset()
+		return
+	}
+	defer s.Close()
+
+	// Hole punch now by forcing a connect
+	pi := peer.AddrInfo{
+		ID:    rp,
+		Addrs: obsDial,
+	}
+	hs.holePunchConnectWithBackoff(pi)
```

Comment:
@Stebalien @vyzo @marten-seemann 

I am a bit concerned about the user experience here. Say `user1` calls `host.Connect`with `user2` to open a `Bitswap` stream later on. Assume both `user1` and `user2` are NATT'd.

Now, the `Host.Connect` call will return as soon as `user1` has a Relayed connection to `user2`. The Hole Punching happens in the background and the user needs to watch out for a direct connection by listening on the `Connected notification` emitted by the Swarm.

Is it worth introducing a `DirectConnect` or similar API on the `Host` that blocks till we have a direct/hole-punched connection with the remote peer and returns a failure if the hole punch fails ?